### DISCUSSION TITLE
Customized BaseSensorOperator fields get reset on reschedule

### DISCUSSION THREAD

**Discussion by manuzhang** [apache/airflow#20414] [Q&A]
Hi all,

I'm using Airflow 2.1.4 and have created a `BaseSensorOperator` whose target time can be specified with `target_time` or `delta`.

```python
class MySensor(BaseSensorOperator):

    @apply_defaults
    def __init__(self, *, target_time=None, delta=None, **kwargs):
        super().__init__(**kwargs)
        self.target_time = target_time
        self.delta = delta

    def poke(self, context):
        if self.target_time is None and self.delta:
            self.target_time = datetime.now() + self.delta

        self.log.info('Checking if the time (%s) has come', self.target_time)
        return datetime.now() > self.target_time
```

When I use `reschedule` mode, the `target_time` is reset each time and hence the `target_time` can never be reached.

```python
t1 = MySensor(
    task_id='my_sensor',
    target_time=None,
    delta=timedelta(days=14),
    poke_interval=10800,
    mode='reschedule',
    dag=dag
)
```

Is this expected? Is there a way to work around it?

**Comment by manuzhang** [comment]
I store the target_time in `Variable` as  a workaround

```
        if self.target_time is None and self.delta:
            ti = context['ti']
            key = f'{ti.dag_id}-{ti.task_id}-target_time'
            target = Variable.get(key, default_var=None)
            if target is None:
                self.target_time = datetime.now() + self.delta
                Variable.set(key, int(self.target_time.timestamp()))
            else:
                self.target_time = datetime.fromtimestamp(int(target))
```

**Reply by potiuk** [reply]
Yep. I am afraid it is expected. In reschedule mode, the operator gets re-created every time it is rescheduled, so you need to make sure to handle this recreation.

One comment here - the `reschedule mode` has always been an 'early access' and \experimental' feature and we are deprecating it (in 2.3 - to be removed in 2.4) in favour of defferable operators introduced in 2.2. 

So I recommend you to migrate to Airlfow 2.2 and switch to defferable operators.

You can read more about early-access/experimental status of Smart Sensors:  https://airflow.apache.org/docs/apache-airflow/stable/concepts/smart-sensors.html and here is the `Defferable Operators' tutorial https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html
